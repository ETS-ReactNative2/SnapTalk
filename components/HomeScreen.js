import React from 'react';
import { Text, View, StyleSheet, FlatList, Image, SnapshotViewIOS } from 'react-native';
import Fire from './Fire';
import _ from "underscore";
import Post from './Post';

export default class HomeScreen extends React.Component{

  state = {
    posts:[],
    isLoading: false,
    postInArray: false,
    result: '',
  };

  componentDidMount(){
    this.getData
  };

  getData = () =>
  {
    this.setState({isLoading:true})
    this.unsubscribe = Fire.shared.firestore
      .collection("posts")
      .get()
      .then(snapshot => {
        
        snapshot.forEach( doc => {
            this.setState({postInArray:false})
            this.state.posts.forEach(currentPost => {

              if (currentPost.postKey == doc.data().postKey) {
                this.setState({postInArray:true})
              }
            })

            if (!this.state.postInArray) {
              this.state.posts.push(doc.data())
            }
        })
      }).finally(()=> this.setState({isLoading:false}))
  }

  renderPost = post => {
    return(
      <Post post = {post}/>
      )
    };
    render() {
      return (
        <View style={styles.container}>
          <View style={styles.header}>
            <Text style={styles.headerTitle}> Feed </Text>
          </View>
          <FlatList 
            style={styles.feed} 
            data={this.state.posts} 
            renderItem={({item}) => this.renderPost(item)} 
            keyExtractor={item => item.id}
            showsVerticalScrollIndicator={false}
            refreshing={this.state.isLoading}
            onRefresh={this.getData}
            /> 
        </View>
      );
    }
  }

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#EFECF4"

  },
  header: {
    paddingTop: 16,
    paddingBottom: 16,
    backgroundColor: "#FFF",
    alignItems: "center",
    justifyContent: "center",
    borderBottomWidth: 1,
    borderBottomColor: "#EBECF4",
    shadowColor: "#454D65",
    shadowOffset: {height: 5},
    shadowRadius: 15,
    shadowOpacity: 0.2,
    zIndex: 10
  },
  headerTitle: {
    fontSize: 20,
    fontWeight: "500"
  },
  feed: {
    marginHorizontal: 16
  },
  feedItem: {
    backgroundColor: "#FFF",
    borderRadius: 5,
    padding: 8,
    flexDirection: "row",
    marginVertical: 8
  },
  avatar: {
    width: 36,
    height: 36,
    borderRadius: 1,
    marginRight: 16
  },
  name: {
    fontSize: 15,
    fontWeight: "500",
    color: "#454D65"
  },
  timestamp: {
    fontSize: 11,
    color: "#C4C6CE",
    marginTop: 4
  },
  post: {
    marginTop: 16,
    fontSize: 14,
    color: "#838899"
  },
  postImage: {
    width: undefined,
    height: 150,
    borderRadius: 5,
    marginVertical: 16
  }
})