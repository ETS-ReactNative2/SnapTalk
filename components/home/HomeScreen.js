import React from 'react';
import { Text, View, FlatList } from 'react-native';
import PostController from '../firebase/PostController';
import Post from '../post/Post';
import Firebase from 'firebase';
import styles from '../home/style/HomeScreenStyle';

export default class HomeScreen extends React.Component {

  state = {
    posts: [],
    isLoading: false,
    isProfileModalVisible: false,
    isDiscoveryTab: this.props.isDiscoveryTab
  };

  componentDidMount() {
    if (this.state.isDiscoveryTab) {
      this.getDiscoveryData
    }
    else {
      this.getHomeData;
    }
  };

  //get the data for the home screen
  getHomeData = async () => {
    const post = await Firebase
      .firestore()
      .collection('users')
      .doc(Firebase.auth().currentUser.uid)
      .get();
    const fieldPathListOfFollowing = new Firebase.firestore.FieldPath(
      'listOfFollowing',
    );
    const arrayOfFollowing = await post.get(fieldPathListOfFollowing);
    this.setState({ posts: [] })
    this.setState({ isLoading: true });
    PostController.shared.firestore
      .collection('posts')
      .get()
      .then(snapshot => {
        snapshot.forEach(doc => {
          if (arrayOfFollowing.includes(doc.data().uid)) {
            this.state.posts.push(doc.data());
          }
        });
        this.state.posts.sort(function (a, b) {
          return parseInt(b.timestamp) - parseInt(a.timestamp);
        });
      })
      .finally(() => this.setState({ isLoading: false }));
  };

  //get the data from the discovery screen
  getDiscoveryData = () => {
    this.setState({ isLoading: true })
    this.setState({ posts: [] })
    this.unsubscribe = PostController.shared.firestore
      .collection("posts")
      .get()
      .then(snapshot => {
        //Fetching posts from database
        snapshot.forEach(doc => {
          this.state.posts.push(doc.data())
        })
        this.setState({ posts: this.state.posts.sort(function (a, b) { return (parseInt(b.timestamp) - parseInt(a.timestamp)) }) })
      }).finally(() => this.setState({ isLoading: false }))
  }

  renderPost = post => {
    return (
      <Post post={post} profilePagePost={false} />
    )
  };

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.header}>
          {this.state.isDiscoveryTab ? <Text style={styles.headerTitle}> Discovery </Text> : <Text style={styles.headerTitle}> Home </Text>}
        </View>
        <FlatList
          style={styles.feed}
          data={this.state.posts}
          renderItem={({ item }) => this.renderPost(item)}
          keyExtractor={item => item.id}
          showsVerticalScrollIndicator={false}
          refreshing={this.state.isLoading}
          onRefresh={this.state.isDiscoveryTab ? this.getDiscoveryData : this.getHomeData}
        />
      </View>
    );
  }
}
