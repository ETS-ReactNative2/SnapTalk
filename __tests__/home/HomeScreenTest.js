import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import HomeScreen from '../../components/home/HomeScreen'

describe('HomeScreen Test', () => {
    let homeScreenSnapshot = renderer.create(<HomeScreen />).toJSON();

    test('HomeTest Render Test', () => {
        expect(homeScreenSnapshot).toMatchSnapshot();
    })
})

