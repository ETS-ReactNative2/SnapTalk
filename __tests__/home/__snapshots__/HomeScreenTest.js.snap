// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Test HomeTest Render Test 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#EFECF4",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "borderBottomColor": "#EBECF4",
        "borderBottomWidth": 1,
        "justifyContent": "center",
        "paddingBottom": 16,
        "paddingTop": 16,
        "shadowColor": "#454D65",
        "shadowOffset": Object {
          "height": 5,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 15,
        "zIndex": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "500",
        }
      }
    >
       Home 
    </Text>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "marginHorizontal": 16,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
